lib/msf/core/db_manager/exploit_attempt.rb

```diff
@@ -27,186 +27,145 @@ def report_exploit_attempt(host, opts)
+  # @return [void]
+  def do_report_failure_or_success(opts)
+    ::ActiveRecord::Base.connection_pool.with_connection {
+      mrefs  = opts.delete(:refs) || return
+      host   = opts.delete(:host)
+      port   = opts.delete(:port)
+      prot   = opts.delete(:proto)
+      svc    = opts.delete(:service)
+      vuln   = opts.delete(:vuln)
+
+      timestamp  = opts.delete(:timestamp)
+      freason    = opts.delete(:fail_reason)
+      fdetail    = opts.delete(:fail_detail)
+      username   = opts.delete(:username)
+      mname      = opts.delete(:module)
+
+      if not vuln
+        # Create a references map from the module list
+        ref_objs = ::Mdm::Ref.where(:name => mrefs.map { |ref|
+          if ref.respond_to?(:ctx_id) and ref.respond_to?(:ctx_val)
+            "#{ref.ctx_id}-#{ref.ctx_val}"
+          else
+            ref.to_s
+          end
+        })
+
+        # Try find a matching vulnerability
+        vuln = find_vuln_by_refs(ref_objs, host, svc)
+      end
 
-    # We have match, lets create a vuln_attempt record
-    if vuln
       attempt_info = {
-        :vuln_id      => vuln.id,
         :attempted_at => timestamp || Time.now.utc,
-        :exploited    => true,
+        :exploited    => (freason.nil? ? true : false),
+        :fail_detail  => fdetail,
+        :fail_reason  => freason,
+        :module       => mname,
         :username     => username  || "unknown",
-        :module       => mname
       }
 
       attempt_info[:session_id] = opts[:session_id] if opts[:session_id]
       attempt_info[:loot_id]    = opts[:loot_id]    if opts[:loot_id]
 
-      vuln.vuln_attempts.create(attempt_info)
+      # We have match, lets create a vuln_attempt record
+      if vuln
+        attempt_info[:vuln_id] = vuln.id
+        vuln.vuln_attempts.create(attempt_info)
 
-      # Correct the vuln's associated service if necessary
-      if svc and vuln.service_id.nil?
-        vuln.service = svc
-        vuln.save
+        # Correct the vuln's associated service if necessary
+        if svc and vuln.service_id.nil?
+          vuln.service = svc
+          vuln.save
+        end
       end
-    end
 
-    # Report an exploit attempt all the same
-    attempt_info = {
-      :attempted_at => timestamp || Time.now.utc,
-      :exploited    => true,
-      :username     => username  || "unknown",
-      :module       => mname
-    }
+      # Report an exploit attempt all the same
 
-    attempt_info[:vuln_id]    = vuln.id           if vuln
-    attempt_info[:session_id] = opts[:session_id] if opts[:session_id]
-    attempt_info[:loot_id]    = opts[:loot_id]    if opts[:loot_id]
+      if svc
+        attempt_info[:port]  = svc.port
+        attempt_info[:proto] = svc.proto
+      end
 
-    if svc
-      attempt_info[:port]  = svc.port
-      attempt_info[:proto] = svc.proto
-    end
+      if port and svc.nil?
+        attempt_info[:port]  = port
+        attempt_info[:proto] = prot || "tcp"
```

Comment:
The use of `"tcp"` here and [here](https://github.com/rapid7/metasploit-framework/pull/5042/files#r27481803) should use the same constant, say `DEFAULT_PROTO`.